src/Dofs/DofHandler2.jl

```diff
@@ -0,0 +1,225 @@
+
+# struct Field
+#     name::Symbol
+#     interpolation::Interpolation
+#     dim::Int
+# end
+
+mutable struct DofHandler{dim,N,T,M}
+    field_names::Vector{Symbol}
+    field_dims::Vector{Int}
+    field_interpolations::Vector{Interpolation}
+    cell_dofs::Vector{Int}
+    cell_dofs_offset::Vector{Int}
+    closed::ScalarWrapper{Bool}
+    grid::Grid{dim,N,T,M}
+end
+
+function DofHandler(grid::Grid)
+    DofHandler(Symbol[], Int[], Interpolation[], Int[], Int[], ScalarWrapper(false), grid)
+end
+
+function Base.show(io::IO, dh::DofHandler)
+    println(io, "DofHandler")
+    println(io, "  Fields:")
+    for i in 1:nfields(dh)
+        println(io, "    ", repr(dh.field_names[i]), " interpolation: ", dh.field_interpolations[i],", dim: ", dh.field_dims[i])
+    end
+    if !isclosed(dh)
+        print(io, "  Not closed!")
+    else
+        println(io, "  Dofs per cell: ", ndofs_per_cell(dh))
+        print(io, "  Total dofs: ", ndofs(dh))
+    end
+end
+
+ndofs(dh::DofHandler) = maximum(dh.cell_dofs) # TODO: This is not very nice
+ndofs_per_cell(dh::DofHandler, cell::Int=1) = dh.cell_dofs_offset[cell+1] - dh.cell_dofs_offset[cell] # TODO: This is not very nice
+isclosed(dh::DofHandler) = dh.closed[]
+nfields(dh::DofHandler) = length(dh.field_names)
+
+function Base.push!(dh::DofHandler, name::Symbol, ip::Interpolation, dim::Int)
+    @assert !isclosed(dh)
+    @assert !in(name, dh.field_names)
+    push!(dh.field_names, name)
+    push!(dh.field_interpolations, ip)
+    push!(dh.field_dims, dim)
+    nothing
+end
+
+# sort and return true (was already sorted) or false (if we had to sort)
+function sortedge(edge::Tuple{Int,Int})
+    a, b = edge
+    a < b ? (return (edge, true)) : (return ((b, a), false))
+end
+
+sortface(face::Tuple{Int,Int}) = minmax(face[1], face[2])
+function sortface(face::Tuple{Int,Int,Int})
+    a, b, c = face
+    b, c = minmax(b, c)
+    a, c = minmax(a, c)
+    a, b = minmax(a, b)
+    return (a, b, c)
+end
+
+# close the DofHandler and distribute all the dofs
+function close!(dh::DofHandler{dim}) where {dim}
+    @assert !isclosed(dh)
+
+    # `vertexdict` keeps track of the visited vertices. We store the global vertex
+    # number and the first dof we added to that vertex.
+    vertexdict = ((Dict{Int,Int}() for _ in 1:nfields(dh))...)
```

Comment:
I guess these should really be vectors since the length of these tuples will be value dependent

Reply:
Yes